packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -247,28 +252,36 @@ export abstract class SharedObjectCore<TEvent extends ISharedObjectEvents = ISha
 	 * {@inheritDoc (ISharedObject:interface).bindToContext}
 	 */
 	public bindToContext(): void {
-		if (this._isBoundToContext) {
-			return;
+		// ensure the method only runs once by removing the implementation
+		// without this the method suffers from re-entrancy issues
+		this.bindToContext = () => {};
+		if (!this._isBoundToContext) {
+			this.runtime.bindChannel(this);
+			// must set after bind channel so isAttached doesn't report true
+			// before binding is complete
+			this.setBoundAndHandleAttach();
 		}
-
-		this._isBoundToContext = true;
-
-		this.runtime.bindChannel(this);
 	}
 
 	/**
 	 * {@inheritDoc @fluidframework/datastore-definitions#(IChannel:interface).connect}
 	 */
 	public connect(services: IChannelServices) {
-		this.services = services;
-		this.attachDeltaHandler();
+		// handle the case where load is called
+		// before connect; loading detached data stores
+		if (this.services === undefined) {
+			this.services = services;
+			this.attachDeltaHandler();
+		}
```

Comment:
Before calling connect() twice was possible with different services. Not sure if valid, or if something broke later at runtime in that case, but now looks like something different would happen to `this.services`?

Reply:
no, it wasn't as `this.services.deltaConnection.attach` throws if called more than once.